Mathlib/LinearAlgebra/RootSystem/Hom.lean

```diff
@@ -104,8 +122,122 @@ lemma Hom.comp_assoc {ι₁ M₁ N₁ ι₂ M₂ N₂ ι₃ M₃ N₃ : Type*} [
     Hom.comp (Hom.comp h g) f = Hom.comp h (Hom.comp g f) := by
   ext <;> simp
 
+/-- The endomorphism monoid of a root pairing. -/
+instance (P : RootPairing ι R M N) : Monoid (Hom P P) where
+  mul := Hom.comp
+  mul_assoc := Hom.comp_assoc
+  one := Hom.id P
+  one_mul := Hom.id_comp P P
+  mul_one := Hom.comp_id P P
+
+@[simp]
+lemma weightMap_one (P : RootPairing ι R M N) :
+    Hom.weightMap (P := P) (Q := P) 1 = LinearMap.id (R := R) (M := M) :=
+  rfl
+
+@[simp]
+lemma coweightMap_one (P : RootPairing ι R M N) :
+    Hom.coweightMap (P := P) (Q := P) 1 = LinearMap.id (R := R) (M := N) :=
+  rfl
+
+@[simp]
+lemma indexEquiv_one (P : RootPairing ι R M N) :
+    Hom.indexEquiv (P := P) (Q := P) 1 = Equiv.refl ι :=
+  rfl
+
+@[simp]
+lemma weightMap_mul (P : RootPairing ι R M N) (x y : Hom P P) :
+    Hom.weightMap (x * y) = Hom.weightMap x ∘ₗ Hom.weightMap y :=
+  rfl
+
+@[simp]
+lemma coweightMap_mul (P : RootPairing ι R M N) (x y : Hom P P) :
+    Hom.coweightMap (x * y) = Hom.coweightMap y ∘ₗ Hom.coweightMap x :=
+  rfl
+
+@[simp]
+lemma indexEquiv_mul (P : RootPairing ι R M N) (x y : Hom P P) :
+    Hom.indexEquiv (x * y) = Hom.indexEquiv x ∘ Hom.indexEquiv y :=
+  rfl
+
+/-- The endomorphism monoid of a root pairing. -/
+abbrev End (P : RootPairing ι R M N) := Hom P P
+
+/-- The automorphism group of a root pairing. -/
+abbrev Aut (P : RootPairing ι R M N) := (End P)ˣ
```

Comment:
I'm not sure this should be the definition of automorphisms. I think instead we should define:
```lean
@[ext]
structure Equiv {ι₂ M₂ N₂ : Type*}
    [AddCommGroup M₂] [Module R M₂] [AddCommGroup N₂] [Module R N₂]
    (P : RootPairing ι R M N) (Q : RootPairing ι₂ R M₂ N₂) where
  /-- A linear equivalence on weight space. -/
  weightMap : M ≃ₗ[R] M₂
  /-- A contravariant linear equivalence on coweight space. -/
  coweightMap : N₂ ≃ₗ[R] N
  /-- A bijection on index sets. -/
  indexEquiv : ι ≃ ι₂
  weight_coweight_transpose : Q.toDualRight ≪≫ₗ weightMap.dualMap = coweightMap ≪≫ₗ P.toDualRight
  root_weightMap : weightMap ∘ P.root = Q.root ∘ indexEquiv
  coroot_coweightMap : coweightMap ∘ Q.coroot = P.coroot ∘ indexEquiv.symm
```
and then use `Equiv P P` for `Aut V` instead of `(End P)ˣ`.

I claim the advantage of using `Equiv P P` is that in due course we will wish to talk about equivalence of root pairings, and these will have to be written using `Equiv P Q`. Life will be simpler if there is one standard way to speak of an equivalence of a root pairing with itself. In other words, if we use `(End P)ˣ` here, since we'll eventually be forced to introduce `Equiv P Q`, we will be forced to author `(End P)ˣ ≃* Equiv P P` and transport things across it. (In fact this design is probably why you need to use `Module.endUnitsToModuleAutEquiv` as part of these changes, though obviously `Module.endUnitsToModuleAutEquiv` is worth having anyway.)

The only annoying thing is a bunch of API will have to be duplicated for `Hom` and `Equiv`, and this will be a tiny bit more awkward than for the analogous situation with `LinearMap` and `LinearEquiv` since we cannot have `Equiv` extend `Hom` (because the data of the maps `M → M₂`, `N → N₂` is not at the top-level in the class fields).

What do you think?

If you agree but are not so keen on writing even more boilerplate, I'd be happy to spend some time on this at the weekend, and could push up a commit.

Reply:
That sounds like an excellent idea. I will try to take care of it myself.

I would like to rename `weightMap` (and similar parts) to something like `weightEquiv` or `weightIsom` in this case.  Does that seem reasonable?

Reply:
I don't have strong feelings on naming here, but the idea of the new names fits nicely with a final thought I had after I wrote the above. Specifically, we could define `RootPairing.Equiv` in a way that would allow us to extend `RootPairing.Hom` (and thus save a little manual API writing) as follows:
```lean
variable {ι₂ M₂ N₂ : Type*}
    [AddCommGroup M₂] [Module R M₂] [AddCommGroup N₂] [Module R N₂]
    (P : RootPairing ι R M N) (Q : RootPairing ι₂ R M₂ N₂)

@[ext]
structure Equiv extends Hom P Q where
    bijective_weightMap : Bijective weightMap
    bijective_coweightMap : Bijective coweightMap

namespace Equiv

def weightEquiv (e : Equiv P Q) : M ≃ₗ[R] M₂ :=
    LinearEquiv.ofBijective _ e.bijective_weightMap

def coweightEquiv (e : Equiv P Q) : N₂ ≃ₗ[R] N :=
    LinearEquiv.ofBijective _ e.bijective_coweightMap

-- More API

end Equiv
```

Reply:
It turns out `Module.endUnitsToModuleAutEquiv` already exists, as `LinearMap.GeneralLinearGroup.generalLinearEquiv`, so I will reconfigure.

Reply:
Okay, I think I put in enough API.  I'm afraid this PR became quite large.